File: core/java-src/eva/Attribute.java
// Copyright 2015-2019 Workiva Inc.
// 
// Licensed under the Eclipse Public License 1.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//      http://opensource.org/licenses/eclipse-1.0.php
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package eva;

import clojure.java.api.Clojure;

/**
 * The Attribute interface provides a programatic representation of a schema attribute.
 *
 * <p>Attributes are loaded into memory, and it is generally more efficient to access
 * attribute information via this interface.
 */
public interface Attribute {
    Object CARDINALITY_MANY = Clojure.read(":db.cardinality/many");
    Object CARDINALITY_ONE = Clojure.read(":db.cardinality/one");
    Object TYPE_BIGDEC = Clojure.read(":db.type/bigdec");
    Object TYPE_BIGINT = Clojure.read(":db.type/bigint");
    Object TYPE_BOOLEAN = Clojure.read(":db.type/boolean");
    Object TYPE_BYTES = Clojure.read(":db.type/bytes");
    Object TYPE_DOUBLE = Clojure.read(":db.type/double");
    Object TYPE_FLOAT = Clojure.read(":db.type/float");
    Object TYPE_FN = Clojure.read(":db.type/fn");
    Object TYPE_INSTANT = Clojure.read(":db.type/instant");
    Object TYPE_KEYWORD = Clojure.read(":db.type/keyword");
    Object TYPE_LONG = Clojure.read(":db.type/long");
    Object TYPE_REF = Clojure.read(":db.type/ref");
    Object TYPE_STRING = Clojure.read(":db.type/string");
    Object TYPE_URI = Clojure.read(":db.type/uri");
    Object TYPE_UUID = Clojure.read(":db.type/uuid");
    Object UNIQUE_IDENTITY = Clojure.read(":db.unique/identity");
    Object UNIQUE_VALUE = Clojure.read(":db.unique/value");

    /**
     * Provides the cardinality of the attribute.
     *
     * @return either {@link #CARDINALITY_ONE} or {@link #CARDINALITY_MANY}
     */
    Object cardinality();

    /**
     * Unsupported.
     *
     * @return true if attribute has :db/fulltext true
     */
    boolean hasFullText();

    /**
     * Unsupported.
     *
     * @return true if attribute has :db/noHistory true
     */
    boolean hasNoHistory();

    /**
     * Entity id of the attribute.
     *
     * @return entity id
     */
    Object id();

    /**
     * Human readable keyword that identifies the attribute.
     *
     * @return keyword identifier
     */
    Object ident();

    /**
     * Tests if the reference attribute is a component-attribute.
     *
     * <p>This indicates that the referenced entity is considered "a part of"
     * the referencing entity.
     *
     * @return true if attribute has :db/isComponent true
     */
    boolean isComponent();

    /**
     * Indicates if the attribute value are indexed.
     *
     * <p>This allows efficient searches by attribute value.
     * By default, all attributes are indexed (except when of type <code>:db.type/bytes</code>)
     *
     * @return true if attribute DOES NOT have :db/index false or :db/valueType :db.type/bytes
     */
    boolean isIndexed();

    /**
     * Indicates if the attribute is in the AVET index.
     * Unlike Datomic, this is always true if isIndexed() is true.
     *
     * @return true if isIndexed() is true
     */
    boolean hasAVET();

    /**
     * Indicates if the attribute is unique, and if so, what type of uniqueness constraint is enforced.
     *
     * @return either {@link #UNIQUE_IDENTITY} or {@link #UNIQUE_VALUE}
     */
    Object unique();


    /**
     * The value type of the attribute.
     *
     * @return one of: {@link #TYPE_BIGDEC} {@link #TYPE_BIGINT} {@link #TYPE_BOOLEAN} {@link #TYPE_BYTES}
     *                 {@link #TYPE_DOUBLE} {@link #TYPE_FLOAT} {@link #TYPE_FN} {@link #TYPE_INSTANT}
     *                 {@link #TYPE_KEYWORD} {@link #TYPE_LONG} {@link #TYPE_REF} {@link #TYPE_STRING}
     *                 {@link #TYPE_URI} {@link #TYPE_UUID}
     */
    Object valueType();

}

File: core/src/eva/attribute.clj
;; Copyright 2015-2019 Workiva Inc.
;; 
;; Licensed under the Eclipse Public License 1.0 (the "License");
;; you may not use this file except in compliance with the License.
;; You may obtain a copy of the License at
;; 
;;      http://opensource.org/licenses/eclipse-1.0.php
;; 
;; Unless required by applicable law or agreed to in writing, software
;; distributed under the License is distributed on an "AS IS" BASIS,
;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
;; See the License for the specific language governing permissions and
;; limitations under the License.

(ns eva.attribute
  (:require [eva.error :as err]
            [schema.core :as s]
            [map-experiments.smart-maps.bijection :refer [bijection]]
            [clojure.string :as cstr]
            [plumbing.core :as pc])
  (:import [eva.Attribute]))

(err/deferror-group attribute-non-resolution
  (:attribute-resolution [:identifier])
  (irresolvable-attribute "Encountered an invalid attribute specification")
  (unresolvable-attribute "This database does not contain the provided attribute"))

(def attr-bij
  (bijection  :db/id          :id
              :db/ident       :ident
              :db/cardinality :cardinality
              :db/valueType   :value-type
              :db/unique      :unique
              :db/noHistory   :no-history
              :db/isComponent :is-component
              :db/fulltext    :fulltext
              :db/indexed     :indexed))

(defprotocol ResolveAttribute
  (resolve-attribute [source identifier] "Resolves and returns the attribute given an identifier. Returns nil if the attribute cannot be resolved.")
  (resolve-attribute-strict [source identifier] "Resolves and returns the attribute given an identifier. Throws if the attribute cannot be resolved."))

(s/defrecord Attribute [id
                        ident
                        cardinality
                        value-type
                        unique
                        no-history
                        is-component
                        fulltext
                        indexed]
  eva.Attribute
  (id [_] id)
  (ident [_] ident)
  (cardinality [_] cardinality)
  (valueType [_] value-type)
  (isComponent [_] (boolean is-component))
  (unique [_] unique)
  (hasAVET [_] true)
  (hasFullText [_] (boolean fulltext))
  (hasNoHistory [_] false)
  (isIndexed [_] (if (some? indexed) (boolean indexed) true)))

(defn id
  ([^Attribute attr] (.id attr))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) (id))))
(defn ident
  ([^Attribute attr] (.ident attr))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) (ident))))
(defn cardinality
  ([^Attribute attr] (.cardinality attr))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) (cardinality))))
(defn value-type
  ([^Attribute attr] (.valueType attr))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) (value-type))))
(defn unique
  ([^Attribute attr] (.unique attr))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) (unique))))
(defn has-avet?
  ([^Attribute attr] (.hasAVET attr))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) has-avet?)))
(defn has-full-text?
  ([^Attribute attr] (.hasFullText attr))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) has-full-text?)))
(defn is-component?
  ([^Attribute attr] (.isComponent attr))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) is-component?)))
(defn ref-attr?
  ([attr] (= :db.type/ref (value-type attr)))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) ref-attr?)))
(defn card-many?
  ([attr] (= :db.cardinality/many (cardinality attr)))
  ([source attr-id] (->> attr-id (resolve-attribute-strict source) card-many?)))

(defn through-attr-bij [a] (get attr-bij a a))
(defn db-attr-map->Attribute* [attr-map]
  (pc/map-keys #(get attr-bij % %) attr-map))

(s/defn db-attr-map->Attribute [attr-map]
  (->> attr-map
       (pc/map-keys #(get attr-bij % %))
       map->Attribute))

(defn reverse-attr?
  "Does the keyword's name start with an underscore ?"
  [kw] (-> kw name (.charAt 0) (= \_)))

(defn rm_
  "Removes '_' from position 0 of keyword names, preserving namespace if defined."
  [kw]
  (assert (= (cstr/starts-with? "_" (name kw))))
  (keyword (namespace kw)
           (-> kw name (.substring 1))))

(defn rm_*
  "idempotent rm_"
  [kw]
  (if (reverse-attr? kw) (rm_ kw) kw))

(defn ref-attr?*
  "ref-attr?, but doesn't care if the attr is reversed with _"
  [db a]
  (or (and (reverse-attr? a)
           (ref-attr? db (rm_ a)))
      (ref-attr? db a)))
